### ISSUE TITLE
Uncaught TypeError: Cannot read property 'bitwarden' of undefined

### ISSUE THREAD

**Issue by Mart124** [bitwarden/browser#1443]
Hi,

## Describe the Bug

I face the following error in the browser's console :
`Uncaught TypeError: Cannot read property 'bitwarden' of undefined`
Leading as a result to broken features in the visited website (unresponsive UI etc...).

Here's the full stack :

![sc](https://user-images.githubusercontent.com/37041094/97719024-e8696480-1ac6-11eb-85b8-23ec643dd748.png)

Important links of the stack are the following ones :
https://webapp4you.eu/js/b2l_scripts.2.js?version=2_62_0
https://webapp4you.eu/js/jquery-1.8.min.js?version=2_62_0

Line `1561` is the following one :
`eval("dataTemp."+x+"='"+addslashes(data[x])+"'"); //V1`

The issue occurs with extension / vault unlocked, even if there's no URI matching / no login saved for this URI.
With extension / vault locked, the issue does not occur.

## Expected Result

We can flawlessly navigate through this web portal with Bitwarden vault unlocked, with or without matching logins, without facing the above issue, so without breaking the UI's features.

## Environment

- Operating system: Windows
- Browser: Firefox, Chrome, Edge (last versions)
- Build Version (go to "Settings" → "About" in the app): 1.46.2

Thx 👍

**Comment by kspearrin** [comment]
Please provide reproduction steps. I am not sure how to get the error you are describing.

**Comment by Mart124** [comment]
Thank you @kspearrin for your response.
I've put the reproduction steps above.
Not sure though I'll be allowed to provide you with test credentials :|

**Comment by Mart124** [comment]
Firefox error message is slightly different, but occurs at the same line :

![firefox](https://user-images.githubusercontent.com/37041094/97992809-7d2fd300-1de3-11eb-9b62-911c238203e7.png)

As if some "bitwarden.com" property (or any other form like "com-bitwarden" etc...) was interfering with the jQuery `data()` function. I did not found how though.

**Comment by Mart124** [comment]
@kspearrin @cscharf any thoughts regarding this weird and blocking bug ?
As a first read, do you think it could be due to a Bitwarden issue (perhaps "too intrusive" in the visited code) ?
Or on the contrary, to the web portal itself ?
Many thanks for your help !

**Comment by cscharf** [comment]
@Mart124 , without access to that site's authenticated portion, it's a bit difficult to reproduce, however looking at the JS code they're using I'll note that our extension *is* manipulating the DOM a little bit by adding attributes for autofill such as `<input .... data-com.bitwarden.browser.user-edited="yes" />`, etc.

It would appear, perhaps, this site's JS code is making some assumptions about it's DOM structure as I see it enumerating data storage and that line in particular is calling `eval()` (which is generally not a good idea in the JavaScript world) and throwing the exception as there's no escaping the building of their `eval()` string either (also probably bad practice in the wild).

Here's what I see from their code:
1. they're subscribing to `change` event on any `.ajaxAutoRequestChange` element, like what's in the form you're referencing
2. then they're pulling in the `.data()` object of that element, `$(this)`, which in the jQuery world gets any attribute value prefixed by `data*` and is stored in a object key subsystem
3. then they fall down to the last `else { ... }` block because that element isn't a date picker or a form element
4. then they're enumerating the object keys of that data object
5. because this would be a type `string`, it's going to `eval()` that using the script's `dataTemp` object (assuming that's used for temporary data, I don't really know) but basically the `eval()` is _assuming_ that the composite key will exist in `dataTemp` object, but it doesn't because we're injecting it as the user manipulates the form
6. `eval()` fails with the given error and ka-boom

This really appears to be an issue with the way that site's scripts are written and the assumptions it makes... The only thing I _could_ think of is for a feature request to disable the Bitwarden extension for certain pages or domains/sites (similar to how pop-up/ad blockers or privacy guards work, etc.) so that way we leave the DOM alone on pages that break otherwise.

**Comment by Mart124** [comment]
First of all, @cscharf, many thanks for your feedback, really appreciated 👍

> the `eval()` is _assuming_ that the composite key will exist in `dataTemp` object

Understood, and bug reproduced, the following will fail :
```
x="com.bitwarden";
eval("dataTemp."+x+"='dummyval'");
```
And the following will succeed :
```
x="com";
eval("dataTemp."+x+"='dummy'");
x="com.bitwarden";
eval("dataTemp."+x+"='val'");
```

> our extension _is_ manipulating the DOM a little bit by adding attributes for autofill such as `<input .... data-com.bitwarden.browser.user-edited="yes" />`

As another possible solution, what about manipulating the DOM for autofill only if the visited URI matches one of the vault logins' URI (only then if the extension is sure to have the opportunity to autofill) ?
We could then set this site's vault login URI to exactly match the login page URI, so that autofill would work, and extension would then not manipulate the DOM / interfere with the rest of the site anymore.

Thank you again 👍

**Comment by cscharf** [comment]
Thanks @Mart124 , I wouldn't be completely opposed to exploring that option, however this is the first we've seen where adding an attribute to certain `<input/>` tags actually causes JavaScript failures like this. We'll have this in our issues list to look into at some point.

**Comment by Mart124** [comment]
Good news, many thanks @cscharf for your feedback.
Of course, feel free for further testing etc... I'm fully available to help, anytime 👍

**Comment by Mart124** [comment]
> As another possible solution, what about manipulating the DOM for autofill only if the visited URI matches one of the vault logins' URI (only then if the extension is sure to have the opportunity to autofill) ?
We could then set this site's vault login URI to exactly match the login page URI, so that autofill would work, and extension would then not manipulate the DOM / interfere with the rest of the site anymore.

I was thinking about this solution again.

This would make Bitwarden compatible with such sites which are let's say sensitive to their DOM structure.
Certainly some already have faced this issue, but without investigating / digging into it (not everyone can).

I also wonder if this could lead to less resources consumption (as browser extension would go through DOM only if needed), thus improving performance at some point, which could be nice for the whole user experience.

Thank you again !

**Comment by Mart124** [comment]
> As another possible solution, what about manipulating the DOM for autofill only if the visited URI matches one of the vault logins' URI (only then if the extension is sure to have the opportunity to autofill) ?
We could then set this site's vault login URI to exactly match the login page URI, so that autofill would work, and extension would then not manipulate the DOM / interfere with the rest of the site anymore.

I was thinking about this solution again.

This would make Bitwarden compatible with such sites which are let's say sensitive to their DOM structure.
Certainly some already have faced this issue, but without investigating / digging into it (not everyone can).

I also wonder if this could lead to less resources consumption (as browser extension would go through DOM only if needed), thus improving performance at some point, which could be nice for the whole user experience.

Technically, could perhaps be rather simple, moving the URL matching mechanism before going through the DOM (if some URL matched), but this is only a quick assumption.

Thank you again !

**Comment by cscharf** [comment]
> Technically, could perhaps be rather simple, moving the URL matching mechanism before going through the DOM (if some URL matched), but this is only a quick assumption.

Thanks! That's not a bad idea, however I do wonder how the mechanism/structure of that would work if the user so choose to auto-fill from another saved credential where the URL didn't match, or, for example, credit card/address info, etc. which don't  necessarily have URLs associated with them, we would still need to have a parsed DOM to identify those fields for autofill (or click-fill); although perhaps the click could explicitly parse the DOM on-demand at that point, would have to look into that more.

**Comment by Mart124** [comment]
> perhaps the click could explicitly parse the DOM on-demand

Yes, that would certainly do the trick, in case it was not already parsed thanks to a matching URL.

> would have to look into that more

For sure, this certainly requires a POC, at least some testing.

Thank you again for keeping an eye on this 👍

**Comment by Mart124** [comment]
Yes, keep this under radar, as seen with @cscharf and @kspearrin above :)
There might be a real improvement here in terms of performance / security / privacy / behavior / compatibility...
Many thanks 👍

**Comment by closebot-bw** [comment]
⚠️ **Stale Issue Notice**

This issue has been automatically marked as stale due to inactivity. It will be closed in **2 weeks** (October 30, 2025) if no further activity occurs.

If this issue is still relevant and you would like to keep it open, please:
- Comment on this issue to show continued interest
- Provide any additional information or updates
- Confirm that the issue still exists in the latest version

Thank you for your contribution to this project! 🙏

**Comment by Mart124** [comment]
Can't reproduce the issue anymore, so closing this.
Thx !
